tutorials/backend/hasura-v3-ts-connector/tutorial-site/content/aggregates.md

```diff
@@ -0,0 +1,183 @@
+
+The NDC spec says that each aggregate should act over the same set of rows that we consider when returning `rows`. That
+is, we should apply any predicates, sorting, pagination, and so on, and then apply the aggregate functions over the
+resulting set of rows.
+ 
+So assuming we have a function called `fetch_aggregates` which builds the SQL in this way, we can fill in the
+`aggregates` in the response:
+
+```typescript
+const aggregates = request.query.aggregates && await fetch_aggregates(state, request);
+```
+
+Now let's start to fill in a `fetch_aggregates` function.
+
+We'll actually copy/paste the `fetch_rows` function and create a new function for handling aggregates. It'd be possible
+to extract that commonality into a shared function, but arguably not worth it, since so much is already extracted out
+into small helper functions anyway.
+
+The first difference is the return type. Instead of `RowFieldValue`, we're going to return a value directly from the
+database, so let's change that to `unknown`.
+
+Next, we want to generate the target list using the requested aggregates, so let's change that.
+
+```typescript
+async function fetch_aggregates(state: State, request: QueryRequest): Promise<{
+    [k: string]: unknown
+}> {
+    const target_list = [];
+
+    for (const aggregateName in request.query.aggregates) {
+        if (Object.prototype.hasOwnProperty.call(request.query.aggregates, aggregateName)) {
+            const aggregate = request.query.aggregates[aggregateName];
+            switch(aggregate.type) {
+                case 'star_count':
+                  
+                case 'column_count':
+                  
+                case 'single_column':
+            }
+        }
+    }
+    
+}
+```
+
+For now, we'll handle the first two cases here, and save the last for when we talk about custom aggregates.
+
+In the first case, we want to generate a target list element which uses the `COUNT` SQL aggregate function.
+
+```typescript
+// ...
+case 'star_count':
+    target_list.push(`COUNT(1) AS ${aggregateName}`);
+    break;
+// ...
+```
+
+In the second case, we'll also use the `COUNT` function, but this time, we're counting non-null values in a single column:
+
+```typescript
+// ...
+case 'column_count':
+    target_list.push(`COUNT(${aggregate.column}) AS ${aggregateName}`);
+    break;
+// ...
+```
+
+We also need to interpret the `distinct` property of the aggregate object, and insert the `DISTINCT` keyword if needed:
+
+```typescript
+// ...
+case 'column_count':
+    target_list.push(`COUNT(${aggregate.distinct ? 'DISTINCT ' : ''}${aggregate.column}) AS ${aggregateName}`);
+    break;
+// ...
+```
+
+Now let's update our generated SQL to use the generated target list:
+
+```typescript
+// ...
+const sql = `SELECT ${target_list.join(", ")} FROM (
+               (
+                 SELECT * FROM ${request.collection} ${where_clause} ${order_by_clause} ${limit_clause} ${offset_clause}
+               )`;
+// ...
+```
```

Comment:
At this point, we don't have a `sql` variable within `fetch_aggregates()`, so a user (read: me) might modify their `sql` variables that's already present and be confused.

```suggestion
We'll create a new generated SQL function within `fetch_aggregates()` to use the generated target list:

```typescript
// ...
const sql = `SELECT ${target_list.join(", ")} FROM (
               (
                 SELECT * FROM ${request.collection} ${where_clause} ${order_by_clause} ${limit_clause} ${offset_clause}
               )`;
// ...
```
```

Reply:
Done